infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java

```diff
@@ -27,18 +31,67 @@ public String getVersion() {
                 // Just for wrap this up
             }
 
-            LogReplicationStreams.VersionString versionStringKey =
-                    LogReplicationStreams.VersionString.newBuilder()
-                            .setName("VERSION").build();
             try (TxnContext txn = corfuStore.txn(NAMESPACE)) {
-                versionString =
-                        ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
-                                versionStringKey).getPayload()).getVersion();
+                getVersion(txn);
                 txn.commit();
             } catch (Exception e) {
                 // Just for wrap this up
             }
         }
         return versionString;
     }
+
+    private String getVersion(TxnContext txn) {
+        LogReplicationStreams.VersionString versionStringKey =
+                LogReplicationStreams.VersionString.newBuilder()
+                        .setName("VERSION").build();
+        versionString =
+                ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
+                        versionStringKey).getPayload()).getVersion();
+        return versionString;
+    }
+
+    /**
+     * @return false if the cluster version is already at my version
```

Comment:
Question, there is no such thing as the cluster version as the code is right now, isn't it? the version stored is the one read from a node. So we should actually take in the change Chris will be working on where we differentiate version per node and check that all are the same..

Reply:
Yes this change depends on Chris's PR, but even then the concept of cluster version is something that will be obtained from the Version Pluging adapter. So you are right that within LR there isn't a cluster version as such but there would be one returned to us from the plugin adapter which we will use for the upgrade and migration decisions.